core/trino-spi/src/main/java/io/trino/spi/block/MapBlockBuilder.java

```diff
@@ -304,16 +303,15 @@ public String toString()
     }
 
     @Override
-    public BlockBuilder newBlockBuilderLike(BlockBuilderStatus blockBuilderStatus)
+    public BlockBuilder newBlockBuilderLike(int expectedEntries, BlockBuilderStatus blockBuilderStatus)
     {
-        int newSize = calculateBlockResetSize(getPositionCount());
         return new MapBlockBuilder(
                 getMapType(),
                 blockBuilderStatus,
-                keyBlockBuilder.newBlockBuilderLike(blockBuilderStatus),
-                valueBlockBuilder.newBlockBuilderLike(blockBuilderStatus),
-                new int[newSize + 1],
-                new boolean[newSize]);
+                keyBlockBuilder.newBlockBuilderLike(expectedEntries, blockBuilderStatus),
```

Comment:
can `keyBlockBuilder`, `valueBlockBuilder` contain less positions than map itself?

Reply:
I think it can (because of nulls)

Reply:
So maybe `valueBlockBuilder.newBlockBuilderLike(blockBuilderStatus)` makes sense here as it was before?

Reply:
on one hand, in the case when there are no nulls (or number of nulls < 20%) `newBlockBuilderLike(blockBuilderStatus)` will create bigger than necessary, on the other hand, if the null rate is close to 100%, `newBlockBuilderLike(expectedEntries, blockBuilderStatus) `will waste some memory if the null rate is the same next time.
Maybe the best solution is to do
`keyBlockBuilder.newBlockBuilderLike(Math.min(expectedEntries , calculateBlockResetSize(keyBlockBuilder.getPositionCount())), blockBuilderStatus)`?

Reply:
Maybe (with a comment)

Reply:
Actually `keyBlockBuilder`, `valueBlockBuilder` can also contain more positions than the map.
`valueBlockBuilder.newBlockBuilderLike(blockBuilderStatus)` seems better.